bauble/src/value.rs

```diff
@@ -2171,40 +2302,251 @@ fn convert_from_copy_with_attributes<'a>(
                 return Err(ConversionError::UnexpectedField {
                     attribute: true,
                     field: ident,
-                    ty: ty_id,
+                    ty: ty_id.value,
                 }
                 .spanned(attribute_span));
             }
 
-            Ok(Val {
+            let val = Val {
                 ty: expected_type,
                 value: value.spanned(span),
                 attributes: attributes.spanned(attribute_span),
-            })
+            };
+
+            if let Some(validation) = types.key_type(expected_type).meta.extra_validation {
+                validation(&val, types).map_err(|err| err.spanned(val.value.span))?;
+            }
+
+            Ok(val)
         }
         BorrowCopyVal::Resolved(val) => {
-            if types.can_infer_from(expected_type, val.ty) {
-                let mut val = val.clone();
-                if let Some(extra) = extra_attributes {
-                    val = set_attributes(val, extra, symbols, add_value, object_name)?;
+            let (is_same, can_infer, instanciable) = (
+                expected_type == val.ty,
+                types.can_infer_from(expected_type, val.ty),
+                types.key_type(expected_type).kind.instanciable(),
+            );
+
+            match (is_same, can_infer, instanciable) {
+                // If the type is the same or can infer and expected type isn't instantiable.
+                (true, true, _) | (false, true, false) => {
+                    let mut val = val.clone();
+                    if let Some(extra) = extra_attributes {
+                        val = set_attributes(val, extra, symbols, add_value, object_name)?;
+                    }
+                    Ok(val)
                 }
-                Ok(val)
-            } else {
-                // Got wrong type in copy.
-                Err(ConversionError::ExpectedExactType {
+                // Expected type is instantiable
+                (false, true, true) => {
```

Comment:
why cannot the type be the same here?

Reply:
We have a match arm for that above

Reply:
I think it's more clear to have false, then you know that this arm only happens in the case where it's `(false, true, true)`